state-recover/clients/smartcontract/src/integrationTest/kotlin/linea/build/staterecover/clients/smartcontract/LineaSubmissionEventsClientIntTest.kt

```diff
@@ -0,0 +1,204 @@
+import tech.pegasys.teku.infrastructure.async.SafeFuture
+import java.util.concurrent.TimeUnit
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.seconds
+import kotlin.time.toJavaDuration
+
+@ExtendWith(VertxExtension::class)
+class LineaSubmissionEventsClientIntTest {
+  private lateinit var contractClient: LineaRollupSmartContractClient
+  private val testDataDir = "testdata/coordinator/prover/v2/"
+
+  // 1-block-per-blob test data has 3 aggregations: 1..7, 8..14, 15..21.
+  // We will upgrade the contract in the middle of 2nd aggregation: 12
+  // shall submit blob 12, stop submission, upgrade the contract and resume with blob 13
+  // val lastSubmittedBlobs = blobs.filter { it.startBlockNumber == 7UL }
+  private lateinit var aggregations: List<Aggregation>
+  private lateinit var blobs: List<BlobRecord>
+  private lateinit var submissionEventsFetcher: LineaRollupSubmissionEventsClient
+
+  private fun setupTest(
+    vertx: Vertx
+  ) {
+    val rollupDeploymentFuture = ContractsManager.get()
+      .deployLineaRollup(numberOfOperators = 2, contractVersion = LineaContractVersion.V6)
+    // load files from FS while smc deploy
+    loadBlobsAndAggregations(
+      blobsResponsesDir = "$testDataDir/compression/responses",
+      aggregationsResponsesDir = "$testDataDir/aggregation/responses"
+    )
+      .let { (blobs, aggregations) ->
+        this.blobs = blobs.sortedBy { it.startBlockNumber }
+        this.aggregations = aggregations.sortedBy { it.startBlockNumber }
+      }
+    // wait smc deployment finishes
+    val rollupDeploymentResult = rollupDeploymentFuture.get()
+    contractClient = rollupDeploymentResult.rollupOperatorClient
+    val eventsFetcherWeb3jClient = Web3jClientManager.buildL1Client(
+      log = LogManager.getLogger("test.clients.l1.events-fetcher"),
+      requestResponseLogLevel = Level.INFO,
+      failuresLogLevel = Level.WARN
+    )
+    submissionEventsFetcher = LineaSubmissionEventsClientWeb3jIpml(
+      logsClient = Web3JLogsClient(
+        vertx = vertx,
+        web3jClient = eventsFetcherWeb3jClient,
+        config = Web3JLogsClient.Config(
+          timeout = 30.seconds,
+          backoffDelay = 1.seconds,
+          lookBackRange = 100
+        )
+      ),
+      smartContractAddress = rollupDeploymentResult.contractAddress,
+      mostRecentBlockTag = BlockParameter.Tag.LATEST
+    )
+  }
+
+  private fun waitTransactionExecution(txHash: String, timeout: Duration = 5.seconds) {
+    await()
+      .atMost(timeout.toJavaDuration())
+      .untilAsserted {
+        val receipt =
+          Web3jClientManager.l1Client
+            .ethGetTransactionReceipt(txHash).send().transactionReceipt
+        assertThat(receipt.isPresent).isTrue()
+      }
+  }
+
+  @Test
+  @Timeout(3, timeUnit = TimeUnit.MINUTES)
+  fun `submission works with contract V6`(
+    vertx: Vertx,
+    testContext: VertxTestContext
+  ) {
+    testSubmission(vertx, testContext)
+  }
+
```

Comment:
Should we also add test cases other than happy path, e.g. when calling the following function with wrong l2StartBlockNumber:
```
submissionEventsFetcher.findDataSubmittedV3EventsUtilNextFinalization(
              l2StartBlockNumberInclusive = expectedFinalizationEvent.startBlockNumber + 1
            )
```
as well as throwing errors from `logsClient`, wdyt?

Reply:
>as well as throwing errors from logsClient, wdyt?

That's a valid concern. Atm will fail. However, I prefer to test resilience/failure later in a holistic rather than just now for the event.